### ISSUE TITLE
radosgw-admin in rook operator 1.12.4 image fails to run commands against Ceph reef

### ISSUE THREAD

**Issue by DaDummy** [rook/rook#13010]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
- Rook operator fails to reconcile OBC after upgrading the cluster to ceph reef. For a new OBC no OB is created.
- `radosgw-admin` when run in a toolbox deployed using the rook operator 1.12.4 image hangs indefinitely for invocations that execute normally in a toolbox utilizing the ceph reef image.

**Expected behavior:**
- Rook operator is capable of reconciling OBC and creating an OB for a newly created OBC.
- `radosgw-admin` when run in a toolbox deployed using the rook operator 1.12.4 image behaves similar to when run from a toolbox utilizing the ceph reef image.

**How to reproduce it (minimal and precise):**
1. Deploy rook v1.12.4
2. Create a cluster running ceph quincy (v17.2.6) with at least one rgw
3. Upgrade cluster to reef (v18.2.0) - issue is probably reproducible when deploying reef directly, but I haven't tested that.
4. Apply [toolbox.yaml](https://github.com/rook/rook/blob/v1.12.4/deploy/examples/toolbox.yaml) and [toolbox-operator-image.yaml](https://github.com/rook/rook/blob/v1.12.4/deploy/examples/toolbox-operator-image.yaml)
5. Run `radosgw-admin user list --cluster="rook-ceph" --conf="/etc/ceph/ceph.conf" --keyring="/etc/ceph/keyring"` or  `radosgw-admin user list` in both toolbox pods and compare.

**Cluster Status to submit**:

```
  cluster:
    id:     ...
    health: HEALTH_OK

  services:
    mon: 3 daemons, quorum h,j,k (age 5h)
    mgr: b(active, since 4h), standbys: a
    mds: 2/2 daemons up, 2 hot standby
    osd: 6 osds: 6 up (since 4d), 6 in (since 3M)
    rgw: 4 daemons active (2 hosts, 4 zones)

  data:
    volumes: 2/2 healthy
    pools:   36 pools, 617 pgs
    objects: 3.12M objects, 9.6 TiB
    usage:   29 TiB used, 29 TiB / 58 TiB avail
    pgs:     615 active+clean
             2   active+clean+scrubbing+deep
```

**Environment**:
* OS (e.g. from /etc/os-release):
   ```
   $ lsb_release -a
   No LSB modules are available.
   Distributor ID: Debian
   Description:    Debian GNU/Linux 12 (bookworm)
   Release:        12
   Codename:       bookworm
   ```
* Kernel (e.g. `uname -a`):
   ```
   $ uname -a
   Linux hostname 6.1.0-12-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.52-1 (2023-09-07) x86_64 GNU/Linux
   ```
* Cloud provider or hardware configuration: Not relevant.
* Rook version (use `rook version` inside of a Rook Pod):
   ```
   $ rook version
   rook: v1.12.4
   go: go1.21.1
   ```
* Storage backend version (e.g. for ceph do `ceph -v`):
   in operator pod:
   ```
   $ ceph -v
   ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
   ```
   in toolbox pod and most others:
   ```
   $ ceph -v
   ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
   ```
* Kubernetes version (use `kubectl version`):
   ```
   $ kubectl version
   Client Version: v1.28.2
   Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
   Server Version: v1.27.5
   ```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Kubernetes on Baremetal deployed using kubespray
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): `HEALTH_OK`

**Comment by DaDummy** [comment]
I currently workaround this issue by running a custom operator image built from `ceph:v18.2.0-20230926`.

I created the image by checking out rook tag v1.12.4 and building the image using the following commandline, then pushing that image to a private docker registry:

```
CEPH_VERSION=v18.2.0-20230926  make -j4 build
```

Please be aware that deploying the operator with an image from a private registry requires setting a pullsecret in all places with the respective comment in [common.yaml](https://github.com/rook/rook/blob/master/deploy/examples/common.yaml) in addition to the operator deployment.

**Comment by travisn** [comment]
We would expect compatibility across the ceph versions so the operator image can be a different ceph version than the ceph daemons. 

The CI runs against the reef image, for example see [this run](https://github.com/rook/rook/actions/runs/6422236541/job/17438331088) on the latest 1.12 branch. The object tests create OBCs and no errors have been detected.

@thotz Can you repro this?

We could potentially update the Rook base image to v1.18, we just hadn't been planning on until Rook v1.13.

**Comment by DaDummy** [comment]
It's possible that repro is more complicated and thereby slipped through the cracks. i.e. the issue might only occur if the  radosgw admin user was created with quincy prior to the upgrade or it might be even more closely tied to my specific setup.

I decided to report it here primarily to communicate that there might be an issue to cautions admins and to communicate the workaround to anyone who might happen to also be affected.

I'm absolutely fine with riding this one out by doing custom builds till the 1.13 release if I turn out to be the only affected party. :)

**Comment by rkachach** [comment]
Not sure if this helps or not but in the past I saw this issue when the RGW gateway is not listening. Is your rgw pod is up&running? is the server listening for connections?

**Comment by DaDummy** [comment]
I had two operator toolboxes running in parallel:

```
$ ceph -v
ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
```
```
$ ceph -v
ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
```

The reef client was able to normally communicate with rgw, while quincy got stuck. Deploying the operator with my custom image based off the ceph reef image did resolve the issue on the operator side as well.

I am positive that this was not a networking issue or a general failure in the rgw.

**Comment by DaDummy** [comment]
> Not sure if this helps or not but in the past I saw this issue when the RGW gateway is not listening. Is your rgw pod is up&running? is the server listening for connections?

To answer more directly:

` ceph -s` is reporting no issues with rgw:

```
  cluster:
    id:     ...
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum h,j,k (age 12h)
    mgr: b(active, since 4h), standbys: a
    mds: 2/2 daemons up, 2 hot standby
    osd: 6 osds: 6 up (since 7d), 6 in (since 3M)
    rgw: 4 daemons active (2 hosts, 4 zones)
 
  data:
    volumes: 2/2 healthy
    pools:   36 pools, 617 pgs
    objects: 3.20M objects, 10 TiB
    usage:   30 TiB used, 28 TiB / 58 TiB avail
    pgs:     564 active+clean
             36  active+clean+scrubbing
             17  active+clean+scrubbing+deep
```

And I could access public files on existing buckets via ingress/http, as well as s3cmd while the issue was occurring. I did not check the individual pods statuses, though. But I did restart them and all hosts at one point. The issue persisted after the restart.

I've deployed my custom operator image, which then worked correctly and finished reconciling. I was still able to reproduce the issue afterwards by following the steps to reproduce from the Issue and [toolbox-operator-image.yaml](https://github.com/rook/rook/blob/v1.12.4/deploy/examples/toolbox-operator-image.yaml).

If needed I can try again. I expect to still be able to reproduce the issue.

**Comment by rkachach** [comment]
@DaDummy I'm not able to reproduce the issue. Basically what I did:

1. I created a test-cluster with rook:v1.12.4 and ceph:v17.2.6
2. created an object store
3. Listed the users successfully 
4. Started an upgrade to ceph:v18.2.0
5. Listed the users successfully 

Below are the logs from my testing cluster. As you can see after the upgrade I can't see the issue. Not sure what's the difference between your env and the one I created. 

```

bash-4.4$ ceph version
ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
bash-4.4$ radosgw-admin user list
[
    "dashboard-admin"
]
bash-4.4$ radosgw-admin user list --cluster="rook-ceph" --conf="/etc/ceph/ceph.conf" --keyring="/etc/ceph/keyring"
[
    "dashboard-admin"
]
bash-4.4$ 
 
bash-4.4$ ceph version
ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
bash-4.4$ 
bash-4.4$ radosgw-admin user list
[
    "dashboard-admin"
]
bash-4.4$ radosgw-admin user list --cluster="rook-ceph" --conf="/etc/ceph/ceph.conf" --keyring="/etc/ceph/keyring"
[
    "dashboard-admin"
]
bash-4.4$ 

```

**Comment by DaDummy** [comment]
@rkachach I am quite positive that the issue is *not* with the rgw pod given that I can reproduce the issue right now with a pod containing radosgw-admin v17.2.6 while interacting normally with rgw at the same time in a second pod that provides radosgw-admin v18.2.0.

Still you are right in that this might be associated with my local environment somehow and that above repro might not be enough. Thanks for looking into it!

**Comment by parth-gr** [comment]
> @rkachach Can you confirm that in the container you were testing with `radosgw-admin` stayed at version 17.2.6?
> 
> ```
> $ ceph version
> ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
> $ radosgw-admin --version
> ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
> ```

I guess version difference make this problem

**Comment by DaDummy** [comment]
This issue occurs for me when using the current rook operator 1.12.5 image. The ceph cluster is upgraded to reef, but the tools shipped with the operator are from quincy as can be seen here:
```
$ ceph version
ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
$ ceph -v
ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
```

**Comment by rkachach** [comment]
In my case the toolbox stayed in 17.2.6 that's because the toolbox.yaml uses `image: quay.io/ceph/ceph:v17.2.6` and I didn't changed the image on the toolbox.yaml. There's no automatic upgrade in this case as the tools are not managed by the operator.

**Comment by rkachach** [comment]
> This issue occurs for me when using the current rook operator 1.12.5 image. The ceph cluster is upgraded to reef, but the tools shipped with the operator are from quincy as can be seen here:
> 
> ```
> $ ceph version
> ceph version 18.2.0 (5dd24139a1eada541a3bc16b6941c5dde975e26d) reef (stable)
> $ ceph -v
> ceph version 17.2.6 (d7ff0d10654d2280e08f1ab989c7cdf3064446a5) quincy (stable)
> ```

You have to update the tools yaml manually to get the desired version.

**Comment by DaDummy** [comment]
@rkachach Alright so you did try using `radosgw-admin` v17.2.6 to access rgw after having the operator upgrade the ceph cluster to v18.2.0 and you did not manage to reproduce the issue with these steps, do I understand that correctly?

Are there any other information I may provide that might help reproducing the issue?

At the moment I still do have a cluster running where `radosgw-admin` v17.2.6 irreducibly gets stuck while `radosgw-admin` v18.2.0 operates normally, so I can look up state information and configs.

**Comment by rkachach** [comment]
@DaDummy yes, your understanding is correct though I saw a similar issue to the one you are describing after some time running the cluster. I mean, at the beginning (just after the upgrade) I was able to access the new cluster running ` v18.2.0` by using the old toolbox `v17.2.6` but after a while it stopped working. 

I understand from your testing that using the right `radosgw-admin`  (with `v18.2.0`) works correctly, right?

**Comment by thotz** [comment]
AFAIR RGW has issues with different versions of radosgw-admin, especially when the lower version of radosgw-admin executes against a newer version of the RGW cluster.  This can be applicable to certain commands, but not all. We had this issue earlier hence Rook internally uses rgw adminops via go-ceph library for operations.

**Comment by DaDummy** [comment]
> @DaDummy yes, your understanding is correct though I saw a similar issue to the one you are describing after some time running the cluster. I mean, at the beginning (just after the upgrade) I was able to access the new cluster running ` v18.2.0` by using the old toolbox `v17.2.6` but after a while it stopped working.
> 
> I understand from your testing that using the right `radosgw-admin` (with `v18.2.0`) works correctly, right?

Thank you for the clarification. That's an interesting development.

Yes, using the ceph `v18.2.0` image as base to build the rook image does make all issues I was observing go away, which translates to using  `radosgw-admin` `v18.2.0` for the issue discussed here.

**Comment by rkachach** [comment]
awesome, since this seems to be a known issue (using older `radosgw-admin` to connect to newer rgw) I'll fix the upgrade docs to clearly indicate the need of upgrading manually the toolbox image as this is not done automatically by the Operator

**Comment by travisn** [comment]
Upgrading the toolbox will be needed, we still need to consider upgrading the Rook operator base image to Reef. 

@thotz If we use Reef as the base image for the operator, it will be able to connect to Pacific or Quincy more reliably? We really need to move all the calls to the go-ceph library to avoid this issues though.

**Comment by thotz** [comment]
> Upgrading the toolbox will be needed, we still need to consider upgrading the Rook operator base image to Reef.
> 
> @thotz If we use Reef as the base image for the operator, it will be able to connect to Pacific or Quincy more reliably? We really need to move all the calls to the go-ceph library to avoid this issues though.

Apart from the multisite site functionality everything uses go-ceph library. At that time go-ceph/admins-ops does not have the capabilities of handling realm, zone, and zonegroup via rest APIs. It is better to change those as well, still not sure whether it is supported or not

**Comment by thotz** [comment]
Go-ceph is safe to use between higher and lower versions

**Comment by DaDummy** [comment]
Sorry to spoil the fun, but while #13050 PR did make a change for which the necessity was discovered here, that PR did not address this issue at all.

Please take note of the issue's title:

> radosgw-admin in rook operator 1.12.4 image fails to run commands against Ceph reef

This issue is explicitly about the **rook operator image** which is not used for the toolbox by default. This issue only uses the toolbox to more easily _demonstrate_ the issue.

I recommend reopening this issue or adding a comment that clearly states that this issue is resolved as `won't fix`.

**Comment by DaDummy** [comment]
> > Upgrading the toolbox will be needed, we still need to consider upgrading the Rook operator base image to Reef.
> > @thotz If we use Reef as the base image for the operator, it will be able to connect to Pacific or Quincy more reliably? We really need to move all the calls to the go-ceph library to avoid this issues though.
> 
> Apart from the multisite site functionality everything uses go-ceph library. At that time go-ceph/admins-ops does not have the capabilities of handling realm, zone, and zonegroup via rest APIs. It is better to change those as well, still not sure whether it is supported or not

OBC user-handling also uses `radosgw-admin`: https://github.com/rook/rook/blob/4b701da916c74282898d8b8fc3dfa32765bd1f2d/pkg/operator/ceph/object/user.go#L159

**Comment by travisn** [comment]
Reopening since #13050 was just for the toolbox, but the issue is with the operator

**Comment by thotz** [comment]
> > > Upgrading the toolbox will be needed, we still need to consider upgrading the Rook operator base image to Reef.
> > > @thotz If we use Reef as the base image for the operator, it will be able to connect to Pacific or Quincy more reliably? We really need to move all the calls to the go-ceph library to avoid this issues though.
> > 
> > 
> > Apart from the multisite site functionality everything uses go-ceph library. At that time go-ceph/admins-ops does not have the capabilities of handling realm, zone, and zonegroup via rest APIs. It is better to change those as well, still not sure whether it is supported or not
> 
> The user-creation/lookup for OBC reconciliation also utilizes `radosgw-admin`:
> 
> https://github.com/rook/rook/blob/4b701da916c74282898d8b8fc3dfa32765bd1f2d/pkg/operator/ceph/object/user.go#L159

This api is used for creating "adminsop" user, not for other users AFAIR. Checking adminsops user happens for all the call, it is resulting in repeatedly calling IMO, better to save the credentials in secret like for external.
